Assorted build/scripts/utils/documentation improvements

PaulSD opened the pull request:
No functional changes to mstpd itself, but lots of changes to the build and runtime scripts...  This should make it much easier for users to get up and running with mstpd.

commodo commented:
Hey

I am traveling at the moment.
Will take a look when back.

Thanks for the contribution:)


On Fri, 13 Jan 2017 at 01:21, Paul Donohue <notifications@github.com> wrote:

> No functional changes to mstpd itself, but lots of changes to the build
> and runtime scripts... This should make it much easier for users to get up
> and running with mstpd.
>
>
>
>
> ------------------------------
>
>
>
>
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/mstpd/mstpd/pull/25
>
>
>
>
> Commit Summary
>
>
>
>
>    - configure.ac: support old versions of automake
>
>    - configure.ac: Adjust the default exec_prefix
>
>    - Move files which include path names to autoconf templates
>
>    - Automatically populate file paths at build time
>
>    - ifupdown.sh.in: fix some formatting inconsistencies
>
>    - Install man pages and bash completion files to the appropriate paths
>
>    - Makefile.am: Improve the names of install destination variables
>
>    - Rename lib/ to utils/ in the source tree
>
>    - ifupdown.sh.in: quote environment variables used as arguments
>
>    - ifupdown.sh.in: use variables for executables
>
>    - ifupdown.sh.in: use 'ip' in place of the old 'ifconfig' command
>
>    - Rename mstpctl-utils.sh to mstpctl-utils-functions.sh
>
>    - mstpctl-utils-functions.sh: remove executable bit and #!
>
>    - Add mstp_config_bridge and ifquery scripts
>
>    - bridge-stp.in: verify the command name of the running PID
>
>    - Add an example bridge-stp.conf file
>
>    - bridge-stp.in: Add a comment about kernel deadlocks
>
>    - bridge-stp.in: automatically start/stop mstpd as needed
>
>    - Configure bridge-stp to manage the mstpd process automatically by
>    default
>
>    - bridge-stp.in: add options to restart and/or reconfigure mstpd
>
>    - Remove the old mstpctl_restart_config script
>
>    - Add a spec file for building RPMs
>
>    - Add files for building Debian packages
>
>    - Document the use of mstpd with VLANs
>
>
>
>
>
>
> File Changes
>
>
>
>
>    -
>
>    *M*
>
>    .gitignore <https://github.com/mstpd/mstpd/pull/25/files#diff-0>
>
>    (18)
>
>
>
>    -
>
>    *M*
>
>    Makefile.am <https://github.com/mstpd/mstpd/pull/25/files#diff-1>
>
>    (115)
>
>
>
>    -
>
>    *A*
>
>    README.VLANs.md <https://github.com/mstpd/mstpd/pull/25/files#diff-2>
>
>    (43)
>
>
>
>    -
>
>    *D*
>
>    bridge-stp <https://github.com/mstpd/mstpd/pull/25/files#diff-3>
>
>    (76)
>
>
>
>    -
>
>    *A*
>
>    bridge-stp.conf <https://github.com/mstpd/mstpd/pull/25/files#diff-4>
>
>    (13)
>
>
>
>    -
>
>    *A*
>
>    bridge-stp.in <https://github.com/mstpd/mstpd/pull/25/files#diff-5>
>
>    (240)
>
>
>
>    -
>
>    *M*
>
>    configure.ac <https://github.com/mstpd/mstpd/pull/25/files#diff-6>
>
>    (23)
>
>
>
>    -
>
>    *A*
>
>    debian/changelog <https://github.com/mstpd/mstpd/pull/25/files#diff-7>
>
>    (5)
>
>
>
>    -
>
>    *A*
>
>    debian/compat <https://github.com/mstpd/mstpd/pull/25/files#diff-8>
>
>    (1)
>
>
>
>    -
>
>    *A*
>
>    debian/control <https://github.com/mstpd/mstpd/pull/25/files#diff-9>
>
>    (22)
>
>
>
>    -
>
>    *A*
>
>    debian/copyright <https://github.com/mstpd/mstpd/pull/25/files#diff-10>
>
>    (53)
>
>
>
>    -
>
>    *A*
>
>    debian/prerm.ex <https://github.com/mstpd/mstpd/pull/25/files#diff-11>
>
>    (56)
>
>
>
>    -
>
>    *A*
>
>    debian/rules <https://github.com/mstpd/mstpd/pull/25/files#diff-12>
>
>    (22)
>
>
>
>    -
>
>    *A*
>
>    debian/source/format
>    <https://github.com/mstpd/mstpd/pull/25/files#diff-13>
>
>    (1)
>
>
>
>    -
>
>    *D*
>
>    lib/ifupdown.sh <https://github.com/mstpd/mstpd/pull/25/files#diff-14>
>
>    (266)
>
>
>
>    -
>
>    *D*
>
>    lib/mstpctl_restart_config
>    <https://github.com/mstpd/mstpd/pull/25/files#diff-15>
>
>    (96)
>
>
>
>    -
>
>    *M*
>
>    main.c <https://github.com/mstpd/mstpd/pull/25/files#diff-16>
>
>    (4)
>
>
>
>    -
>
>    *A*
>
>    rpm.spec <https://github.com/mstpd/mstpd/pull/25/files#diff-17>
>
>    (111)
>
>
>
>    -
>
>    *R*
>
>    utils/bash_completion
>    <https://github.com/mstpd/mstpd/pull/25/files#diff-18>
>
>    (0)
>
>
>
>    -
>
>    *A*
>
>    utils/ifquery <https://github.com/mstpd/mstpd/pull/25/files#diff-19>
>
>    (32)
>
>
>
>    -
>
>    *A*
>
>    utils/ifupdown.sh.in
>    <https://github.com/mstpd/mstpd/pull/25/files#diff-20>
>
>    (235)
>
>
>
>    -
>
>    *A*
>
>    utils/mstp_config_bridge.in
>    <https://github.com/mstpd/mstpd/pull/25/files#diff-21>
>
>    (88)
>
>
>
>    -
>
>    *R*
>
>    utils/mstpctl-utils-functions.sh
>    <https://github.com/mstpd/mstpd/pull/25/files#diff-22>
>
>    (2)
>
>
>
>    -
>
>    *R*
>
>    utils/mstpctl-utils-interfaces.5
>    <https://github.com/mstpd/mstpd/pull/25/files#diff-23>
>
>    (0)
>
>
>
>    -
>
>    *R*
>
>    utils/mstpctl.8 <https://github.com/mstpd/mstpd/pull/25/files#diff-24>
>
>    (0)
>
>
>
>
>
>
>
>
> Patch Links:
>
>
>
>
>    - https://github.com/mstpd/mstpd/pull/25.patch
>
>    - https://github.com/mstpd/mstpd/pull/25.diff
>
>
>
>
>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mstpd/mstpd/pull/25>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACo3jIAX2PZoLXAgFkbl2jbrsqZlYmnVks5rRsNsgaJpZM4LiZI0>
> .
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>
>

commodo commented on utils/ifupdown.sh.in:
```diff
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/bin/bash
```
i'd prefer to keep `/bin/sh` ; because it can be symlinked to `/bin/bash or other shells
some distros use other shells as defaults

PaulSD replied:
The problem here is that I added a bashism: ${IF_MSTPCTL_MAXWAIT//[0-9]/}
${x//x/x} only works in bash and does not work in plain sh or other shells that emulate plain sh.
So, either the shebang needs to be changed to explicitly use bash, or we need to find another way to verify whether $IF_MSTPCTL_MAXWAIT is a valid number.

vivien replied:
I would prefer to write a POSIX shell script. It's not common to have bash on an embedded system.

PaulSD replied:
I can see if I can come up with a POSIX way to do this ... However, note that this script probably would not be used on an embedded system.  This script is intended to be used by ifupdown, which is a Debian/Ubuntu specific package, and bash is considered a required package on Debian/Ubuntu, so it should always be present on any system with ifupdown.  I did add new mstp_config_bridge and ifquery scripts in a separate commit to allow this script to be used on non-Debian/Ubuntu systems, but mstp_config_bridge also requires bash, and ifquery requires python, so again they are unlikely to be used on embedded systems.

PaulSD replied:
Also note that bridge-stp (which is more likely to be used on an embedded system) already contains several bashisms and uses #!/bin/bash

commodo replied:
ah ; then we'll need to address that [ possibly later ]

vivien replied:
> This script is intended to be used by ifupdown, which is a Debian/Ubuntu specific package

A buildsystem such as buildroot provides an ifupdown package which doesn't require Bash.

> Also note that bridge-stp (which is more likely to be used on an embedded system) already contains several bashisms and uses #!/bin/bash

You are correct, this was on my todo list of stuffs to fix. I'll open a ticket.

PaulSD replied:
I'll work on removing the bashisms now... :)

commodo commented:
@PaulSD 
Again ; thanks for the contribution.
I've quickly gone through the commits.

Overall the changes are good [ I did not read too in-depth through all of them ].
Some of changes/renames seem to do more with [your] personal preference.

And I need to come back and check whether all changes are consistent [ in this PR ].
I had the feeling some parts are being changed a few times [across this PR], like you were changing your mind.
[Maybe I'm wrong ; I'm a bit slow from the travel]

Let's do it like this: I'll start picking out some sets of commits that can go in first, and try to group the other changes in a way I can follow better.
That way, things will get moving ; and the number of commits to review will decrease.

Do you prefer that you split this PR into multiple smaller PRs ; I can recommend/ack which commits could be the first batch ?
Or is it ok if I push directly to master the commits [ with you as author ] ?
Your choice.

Thanks
Alex

PaulSD commented:
These commits were not written as a chronological series of changes.  I actually made all my changes without committing anything, then went back through the giant patch and tried to break it up into lots of smaller logically-separate commits that I thought would be easier to explain/review.  I also tried to break out personal preferences (like renames) as separate commits so they could be removed more easily if necessary.  You will definitely see some parts being changed multiple times ... This is not because I changed my mind, but rather because the same piece of code is affected by multiple not-directly-related changes, and I wanted to ensure that the code would be functional even if one of those changes was not applied.

Feel free to push individual commits directly to master.  You may find that some later commits will not apply properly if you skip some earlier commits ... Feel free to fix the merge conflicts yourself, or you can give me a list of commits you would like to apply, and I can fix the merge conflicts and generate new PRs.

commodo commented on utils/mstpctl-utils-interfaces.5:
```diff
@@ -53,8 +53,8 @@ iface br0 inet static
     mstpctl_portrestrtcn swp1=no swp2=no
-    mstpctl_treeprio 32768
-    mstpctl_treeportprio swp3=128
+    mstpctl_treeprio 8
```
let's make this commit part of an initial set of commits;
the patch to `rpm.spec` may move to commit https://github.com/mstpd/mstpd/pull/25/commits/9e9639fa662e011a6d3970f04275760b9be9eeb0

the rest would be handled by the rename commits

commodo commented on configure.ac:
```diff
@@ -6,7 +6,7 @@ AM_INIT_AUTOMAKE([foreign -Wall])
 
 AC_PROG_CC
-AM_PROG_AR
+m4_ifdef([AM_PROG_AR], [AM_PROG_AR])
```
this could be part of an initial commit-set

commodo commented on .gitignore:
```diff
@@ -5,8 +5,8 @@ mstpd
 
 /bridge-stp
 /bridge-stp.tmp
-/lib/ifupdown.sh
```
this rename of `lib` ==> `utils` i'd drop
it's a preference change
it could be that we'd later get a preference change [from someone else] to rename back, or to some other folder name

so, rather than going down that path, i'd prefer to drop this

commodo replied:
ah ; i forgot to add ; if you do really want this change, there is always the democratic approach, where other members of the group could be persuaded otherwise ; 
so, if they are convinced by it, i am fine with it

PaulSD replied:
My thinking here was that lib/ is typically used either to store library dependencies required by the project or to store source files used to build a library provided by the project, so having shell scripts, man pages, etc in there doesn't line up with the source directory layout I've come to expect from most projects, and I was a bit confused when I first looked for those scripts in the source and found them under the lib/ directory.  I used utils/ as the new name because most of the files in this directory seemed to be derived from brctl-utils, but that choice of a new name was rather arbitrary - basically anything other than lib/ would have eliminated my confusion here.

However, I don't feel particularly strongly about this ... If you don't agree that lib/ is a confusing name for this directory, then I'm fine with dropping this commit.

commodo replied:
@vivien @dvitas @asatish 
any thoughts/votes on this ?

i'm neutral

vivien replied:
I tend to be neutral too about this, but I can understand the confusion. "lib" (as well as "src") indeed sounds like an internal component, while "utils", "tools" and such sound like scripts, wrappers, mostly optional helpers.

As this commit is properly isolated (thanks Paul, hence can be easily reverted), I'm fine with keeping the change.

commodo replied:
alright ; let's remove the neutrality from this ; 
let's make it an ack from me (2 neutrals should convert to ack, in this case)

asatish replied:
ack for the change.

commodo commented on README.VLANs.md:
```diff
@@ -0,0 +1,46 @@
+Using mstpd with Trunk Ports / VLANs
```
this file could go as part of an initial commit set
it would be then added to `rpm.spec` when the entire `rpm.spec` is added

commodo commented on configure.ac:
```diff
@@ -1,12 +1,33 @@
+# AC_PREFIX_DEFAULT(...) can be used to override the default prefix, but there
+# doesn't seem to be any standard way to override the default exec_prefix, so
+# just do it manually.
+test "x$exec_prefix" = xNONE && exec_prefix=''
```
so, it doesn't work to call `./configure --exec-prefix= --prefix=` ?
I always thought that was enough to set the prefix values to whatever was needed

if this is needed for the RPM/DEB creation, I'd move the arguments there, and not override them here

PaulSD replied:
Yes, `./configure --exec-prefix=/sbin/` works.  However, prior to this commit, running `./configure` without any arguments will cause bridge-stp to be installed as /usr/local/sbin/bridge-stp ... Which means the usual `./autogen.sh ; ./configure ; make ; sudo make install` will not do what the user expects.  With this commit, `./configure` without any arguments will cause bridge-stp to be installed as /sbin/bridge-stp ... although the user can still override this using `./configure --exec-prefix=...` if they want.

commodo replied:
hmm ; ok
that's actually a good point [ about `/sbin/bridge-stp` ]
let's have this as part of the initial commit set

i think all changes to `configure.ac` could be part of the initial commit set

commodo commented:
@PaulSD 
I'll stop looking over stuff for today.

If it's fine with you, you can create a branch with the commits I proposed for merging and spin up a PR.
I hope you made a backup branch for this ; to check [at the end of it], that you did not miss anything.

I think getting through all changes would take a few days.
But we will get through all of them eventually.

PaulSD commented:
Sounds good.  Thanks for reviewing!

PaulSD commented:
I've opened PR #27 for the first set of commits.

I've also updated this PR:
- Rebased on top of PR #27 
- Moved renames to the end of the series, so that other commits can be applied without them
- Replaced bashisms with POSIX compatible equivalents and changed shebangs to /bin/sh
- Various other reorganization and minor improvements based on comments and things I noticed when I reviewed the commits again

commodo commented on lib/ifupdown.sh.in:
```diff
@@ -1,11 +1,15 @@
 #!/bin/sh
 
+# This script should be symlinked into /etc/network/ip-pre-up.d/ and
```
@PaulSD 
out of curiosity ; 
are you using some sort of automatic formatting tool, or is this manual ?

if automatic, i'd like to try to include it in the Travis CI build ;

PaulSD replied:
I'm doing this manually.  There are some automatic formatting tools out there, but I'm not using one.

commodo replied:
regarding this commit ; i've been undecided ; 
i was also trying to find an automatic formatting tool for shell scripts ; 
and trying to find a style-book [or reference] for shells ; did not find anything

i only found https://www.shellcheck.net/ ;  which checks for potential issues with scripts [ but does not do any formatting ]

so, regarding this change, I'll say ack ;
and we can make this the preferred style for shell scripts for next commits [ that touch on shell scripts ]

then later, I'll probably add stylecheck to do a check on scripts ;

PaulSD replied:
Prior to this commit, ifupdown.sh had a few blocks that were formatted like bridge-stp, and had a lot of blocks that were formatted differently.  My primary goal with this commit was to make the ifupdown.sh formatting internally consistent - one way or the other.  Since bridge-stp already seemed to have consistent formatting, I thought it would be best to mimic that.  This formatting is actually not exactly the formatting I personally prefer to use in shell scripts ... so if you want to change the "preferred style" to something else, that is fine with me.  I was just trying to make it consistent.

There is a formatter here, although I haven't used it enough to say how well it works in practice: http://arachnoid.com/python/beautify_bash_program.html

commodo replied:
i also don't have any preferred style for shell scripts ; 
but we will use this as the project's preferred style ; 
i might have to write a style doc ;

will also check out the ` http://arachnoid.com/python/beautify_bash_program.html` tool

thanks

commodo commented on Makefile.am:
```diff
@@ -16,6 +16,20 @@ if ENABLE_DEVEL
 endif
 mstpctl_CFLAGS = $(mstpd_CFLAGS)
 
+all-local: bridge-stp lib/ifupdown.sh
```
ack for this, for next PR

commodo commented on Makefile.am:
```diff
@@ -18,9 +18,22 @@ mstpctl_CFLAGS = $(mstpd_CFLAGS)
 
 all-local: bridge-stp lib/ifupdown.sh
 
+utilsdir=$(libexecdir)/mstpctl-utils
```
ack this for next PR

commodo commented on Makefile.am:
```diff
@@ -46,22 +46,26 @@ lib/ifupdown.sh: $(srcdir)/lib/ifupdown.sh.in
+	install -m 755 lib/mstpctl-utils.sh $(utilsdest)
+	install -m 755 lib/mstpctl_restart_config $(utilsdest)
+	if [ -d $(etcdest)/network/if-pre-up.d ] ; then ln -sf $(utilsdir)/ifupdown.sh $(etcdest)/network/if-pre-up.d/mstpctl ; fi
+	if [ -d $(etcdest)/network/if-post-down.d ] ; then ln -sf $(utilsdir)/ifupdown.sh $(etcdest)/network/if-post-down.d/mstpctl ; fi
```
you are reverting `utilsdest` to `utilsdir` here
otherwise ; ack

PaulSD replied:
The reversion from utilsdest to utilsdir is actually intentional - the original use of $DESTDIR in the `ln` command is a bug ... I probably should have made this change a separate commit of its own to specifically identify it as a bug I was fixing...

In general, $DESTDIR is intended to facilitate the installation of the compiled software on a different system.  When compiling, $DESTDIR is not used, and any absolute path names (ld library paths, etc) that are compiled into the generated files do not include $DESTDIR.  When installing, $DESTDIR is used, but only to redirect the location of the installed files so that they can be packaged up and moved to another machine without actually installing them in real paths on the local machine; absolute paths compiled into the installed files should not be changed to include $DESTDIR.  After $DESTDIR is packaged and moved to the other machine, the idea is that it should be possible to extract the package into / on the target machine and everything will work as if `make install` (without DESTDIR) was run directly on the target machine.

If we use `ln -sf $(utilsdest)/ifupdown.sh $(etcdest)/network/if-pre-up.d/mstpctl` in the Makefile, then a valid symlink will be created under DESTDIR, but this will become a broken symlink after it is moved to the target machine: `/etc/network/if-pre-up.d/mstpctl -> /destdir/lib/mstpctl-utils/ifupdown.sh`.  We should instead use `ln -sf $(utilsdir)/ifupdown.sh $(etcdest)/network/if-pre-up.d/mstpctl` which creates a broken symlink on the local machine but causes a valid symlink to be installed on the target machine: `/etc/network/if-pre-up.d/mstpctl -> /lib/mstpctl-utils/ifupdown.sh`.

commodo commented on Makefile.am:
```diff
@@ -55,7 +55,7 @@ install-data-hook:
 	install -m 755 bridge-stp $(sbindest)/bridge-stp
 	mkdir -pv $(utilsdest)
 	install -m 755 lib/ifupdown.sh $(utilsdest)
-	install -m 755 lib/mstpctl-utils.sh $(utilsdest)
```
ack for next PR

commodo commented:
@PaulSD 
i've acked a few more commits ; 
please be aware [ if you are, disregard this line ] that if you push stuff to your `master` branch, they show up in this PR 

can you rebase this branch and force-push ?
i like to see the number of commits going down
it gives me a sense of progress, and it's easier to track 

thanks

commodo commented:
@PaulSD 
hey ; don't know if i'll have time today to look over some commits ; will make some time tomorrow 
my day is ending here

PaulSD commented:
No worries.  I will be traveling tomorrow thru 1/23, so I may be slow to respond, but I will try to stay on top of this as you make further progress.

Thanks!

commodo commented on lib/ifupdown.sh.in:
```diff
@@ -16,209 +16,209 @@ fi
 
 all_interfaces= &&
 unset all_interfaces &&
 mstpctl_parse_ports $INTERFACES | while read i; do
```
shellcheck.com says you missed quotes on the `$INTERFACES` var

it also says other stuff, but this is the only one that fits for this commit

commodo replied:
also, ack for next PR for this

commodo replied:
btw; this comment above [with shellcheck.com] was not intended to be snarky
sorry if it came out like that

PaulSD replied:
No worries, all good :)

mstpctl_parse_ports expects the shell to word-split `$INTERFACES` into multiple arguments, so this actually won't work right if `$INTERFACES` is quoted.  Similar issue in some of the `for` loops further down in the script ... for example, `for portpathcost in $portpathcosts; do` won't work right if `$portpathcosts` is quoted, because the for loop requires the shell to word-split `$portpathcosts`.

commodo commented on lib/ifupdown.sh.in:
```diff
@@ -9,11 +9,22 @@
 #
 # Author: Satish Ashok, <sashok@cumulusnetworks.com>
 
-if [ ! -x '@mstpctlfile@' ]; then
```
ack

commodo commented on lib/ifupdown.sh.in:
```diff
@@ -18,9 +18,9 @@ if [ -z "$brctl" ] || [ ! -x "$brctl" ]; then
-ifconfig="$(which ifconfig)"
-if [ -z "$ifconfig" ] || [ ! -x "$ifconfig" ]; then
-  echo "'ifconfig' binary does not exist or is not executable" >&2
+ip="$(which ip)"
```
i was trying to think if there are cases where `ifconfig` would still be needed, but `ip` has been around for a while

ack

commodo commented on lib/ifupdown.sh.in:
```diff
@@ -227,6 +227,66 @@ if [ "$MODE" = 'start' ]; then
+      echo "Timed out waiting for STP on bridge '$IFACE' to converge" >&2
+    fi
+  fi
+
```
ack for this  too

for arithmetic ops, you can drop [if you want] the $ signs
https://github.com/koalaman/shellcheck/wiki/SC2004

so, `count=$(($count+1))` ==> `count=$((count+1))`

PaulSD replied:
Good catch!  I will fix this.

commodo commented on lib/ifupdown.sh.in:
```diff
@@ -29,10 +40,10 @@ esac
 # Previous work (create the interface)
 if [ "$MODE" = 'start' ] && [ ! -d "/sys/class/net/$IFACE" ]; then
-  brctl addbr "$IFACE" || exit 1
+  "$brctl" addbr "$IFACE" || exit 2
```
curious: why `exit 2` ?
i was not gonna ask initially, since `2` seems to be an error code implying a higher error state

but in commit 58468163ac6f320fafe80fe0168c37624cf8e470  you're using `exit 1` in `lib/mstp_config_bridge.in `

PaulSD replied:
Another good catch!  I actually intended to use `exit 2` in commit 5846816 ... but somewhere along the way I lost that.  I will fix this too.

I typically use `exit 1` for generic "soft" errors that can be corrected by running the program differently: bad arguments, bad environment, etc.  I typically use `exit 2` for generic "hard" errors that cannot be corrected simply by running the program differently: missing installation files, missing kernel support, etc.  This distinction might be useful if someone is actually checking the exit codes from the script.

commodo commented on Makefile.am:
```diff
@@ -25,7 +25,7 @@ mstpctlfile=$(sbindir)/mstpctl
 bridgestpconffile=$(sysconfdir)/bridge-stp.conf
 ifupdownfile=$(utilsdir)/ifupdown.sh
-utilsfuncfile=$(utilsdir)/mstpctl-utils.sh
+utilsfuncfile=$(utilsdir)/mstpctl-utils-functions.sh
```
ack

commodo commented:
@PaulSD 
managed to go through a few commits

Thanks

PaulSD commented:
Added PR #29 and updated the remaining commits in this PR per comments.

Thanks!

commodo commented on lib/ifupdown.sh.in:
```diff
@@ -48,6 +48,8 @@ fi
 
 all_interfaces= &&
 unset all_interfaces &&
+# $INTERFACES should be word split into multiple arguments to
```
ack for next PR

commodo commented on lib/ifquery:
```diff
@@ -0,0 +1,32 @@
+def parse_config(interfaces_file, interface):
+    f = open(interfaces_file)
+    intf = ''
+    for line in f:
```
[this may be me, not remembering Python correctly], but does this work ? 
i usually would do  `f.readlines()` or something of the sorts

PaulSD replied:
Yes, this does work, see: https://docs.python.org/2/tutorial/inputoutput.html#methods-of-file-objects
f.readlines() would also work, but that requires more memory since it loads the whole file into a list before looping over it, as opposed to loading the file one line at a time.

commodo commented on lib/ifquery:
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/python
+# Crudely emulate ifquery on systems that don't have it.
```
i must admit, getting rid of some of the Python code, makes me a bit greedy about wanting to get rid of all of it;
that would also mean, that I'd try to replace `ifquery` with shell [ simply to reduce the dependency on Python ]

but let's leave it as is for now

commodo commented on lib/mstp_config_bridge.in:
```diff
@@ -0,0 +1,85 @@
+mstpctl='@mstpctlfile@'
+if [ ! -x "$mstpctl" ]; then
+  echo "'mstpctl' binary does not exist or is not executable" >&2
+  exit 2
```
just to make sure ; `exit 2` [here] is intended, right ?

PaulSD replied:
Yes, intended.

commodo commented on lib/mstp_config_bridge.in:
```diff
@@ -0,0 +1,85 @@
+done
+)"
+
+# Call the mstp ifupdown.sh script.
```
rest looks fine
after addressing my comments, it's ack from me

commodo commented on bridge-stp.in:
```diff
@@ -18,9 +21,9 @@ conf_file='@bridgestpconffile@'
-# Set this to the list of bridges for which MSTP should be used.
-# Set to empty value to allow any valid bridge to use MSTP.
-MSTP_BRIDGES="br0"
+# A space-separated list of bridges for which MSTP should be used in place of
```
ack

commodo commented on bridge-stp.in:
```diff
@@ -35,7 +35,7 @@ checkpid()
         for p in $line; do
-            if [ -z "${p//[0-9]/}" ] && [ -d "/proc/$p" ]; then
+            if [ "$p" = "${p%[!0-9]*}" ] && [ -d "/proc/$p" ]; then
                 pid="$pid $p"
```
ack

commodo commented on bridge-stp.in:
```diff
@@ -2,6 +2,10 @@
 #
+# To avoid kernel deadlocks, this script and any foreground processes it runs
+# must not make any changes (using brctl, ifconfig, ip, /sys/..., etc) to the
+# bridge or any associated kernel network interfaces.
```
ack

commodo commented on bridge-stp.in:
```diff
@@ -52,7 +58,7 @@ case $2 in
             . "$conf_file" || exit
         fi
-        checkpid $pid_file || exit 1
+        checkpid || exit 1
```
personally, i like having the `pid_file` as argument here
but this is more of a preference

commodo replied:
a general question: isn't `/var/run` mounted as tmpfs, or doesn't it get cleaned up on reboot ?

it would seem that that would be a better solution to handle unclean reboots, than doing special pid handling

i agree there's the segfault case [ or some other mstpd crash type ]
in that case I'd also need to check how to do proper crash-handling

PaulSD replied:
I removed the argument because the original `checkpid` code was ignoring it when it actually read the PID file.  I will put it back and fix the code so it uses the argument.

PaulSD replied:
`/var/run` is cleaned up on reboot on many systems.  However, it is not guaranteed to be cleaned up on reboot.  For example, I have a number of RHEL5 and RHEL6 systems that don't clean it up on reboot (It is possible my systems are misconfigured somehow, but I /var/run is not mounted as tmpfs, I don't see any startup scripts to clean it up, and I've confirmed that it is not cleaned up on reboot).

As you mentioned, segfaults are another case we need to handle gracefully.  I don't think there is any "proper" way to handle PID files after a crash.

So, unless there is a better way to verify that an arbitrary process is really an mstpd process, I think this additional PID checking is necessary...

commodo commented on bridge-stp.in:
```diff
@@ -25,18 +25,24 @@ net_dir=/sys/class/net
+                # Make sure a process with this PID is running, that its command
+                # line contains 'mstpd', and that it is not the grep command we
+                # are using to look for the 'mstpd' string.
+                if ps -o command -ww "$p" | grep -v grep | grep 'mstpd' >/dev/null 2>&1; then
```
why not just use the `pidof` function ?

PaulSD replied:
`pidof` is not a function.  It is a binary that is typically included with sysvinit-utils.  It should be installed on most systems, but it may not always be installed.

I can switch this to use `pidof` if you would prefer that.

PaulSD replied:
Actually, I guess using `pidof` allows us to avoid the use of a PID file entirely.  I like that.

commodo commented on bridge-stp.conf:
```diff
@@ -1,7 +1,7 @@
 # needed.  If 'n', mstpd will not be automatically started/stopped by the
 # bridge-stp script.
-MANAGE_MSTPD='n'
+#MANAGE_MSTPD='n'
```
squash this commit in  commit 9ac71b1a6f4a4d3f416d5f8a308aa9355c4cd141 ?
it looks a bit redundant to me to add it there, and disable it here

PaulSD replied:
91c71b1 adds new functionality but doesn't change the default behavior.  This commit changes the default behavior.  I separated it out in case we wanted to skip changing the default behavior, or in case we wanted to revert the default later.  But I can merge it with 91c71b1 if you want.

commodo commented on bridge-stp.in:
```diff
@@ -59,27 +114,65 @@ checkpid()
+        # Determine whether the specified bridge should use MSTP.
+        if ! is_mstp_bridge "$bridge"; then
+            echo "Ignoring bridge '$bridge' that is not listed in \$MSTP_BRIDGES"
+            exit 10
```
`exit 10`  ? 
[curios] how should this be treated on exit ?
or does it have some special meaning ?

PaulSD replied:
The kernel only cares whether the exit status is zero or non-zero, so `exit 1`, `exit 2`, `exit 10`, etc are all treated the same by the kernel.  However, for debugging purposes, newer kernels will print the exit status of the script so the user can figure out what is going on.

In this case, if a bridge is missing from `$MSTP_BRIDGES`, that's not really an "error" ... it is a valid configuration that is intentionally causing a non-zero error status to be returned to the kernel so that this bridge will use kernel STP instead of mstpd.  In that case, I want the exit status printed by the kernel to be clearly different than the exit status which is printed when there is a real error.  Hence why I used 1/2/3 for real errors and 10 for this "error".  There is nothing special about 10 except that it is obviously different than 1/2/3.

commodo commented on bridge-stp.in:
```diff
@@ -20,15 +20,70 @@ if [ $# -ne 2 ]; then
+{
+    Out="$("$mstpctl" showbridge 2>&1)" && return 0
+    # if [ "$Out" contains "Couldn't connect to server" ]; then
+    if [ "$Out" != "${Out%Couldn\'t connect to server*}" ]; then
```
this looks a bit brittle ; i'm not sure how well this handles [long-term] the check `Couldn't connect to server`
i admit I don't have a better idea atm ; will try to think of something

PaulSD replied:
Perhaps we could update mstpctl so it returns a different exit status if it couldn't connect vs if another error occurred?

commodo replied:
yep ; that makes sense to use a special return code from mstpctl to if it couldn't connect;

commodo commented on bridge-stp.in:
```diff
@@ -59,27 +114,65 @@ checkpid()
+        fi
+
+        # Determine whether the specified bridge should use MSTP.
+        if ! is_mstp_bridge "$bridge"; then
```
i like moving this to it's own function ; 
can we make this it's own commit [ preceding this one ] ?
i'd also move lines 117-121 [ from above ] into that function

PaulSD replied:
Ok, will do that and update this PR shortly.

commodo commented:
@PaulSD curios: is it possible to move the .deb & .rpm creation build rules into separate PR at this point in time ?
or do they require some commits before doing that ?

PaulSD commented:
Commit 1b8bb09 is needed for the .deb and .rpm creation stuff ... However, after PR 30 is merged, I can move the .deb & .rpm creation into a separate PR.

PaulSD commented:
I've reworked the remaining commits to use pidof to detect whether mstpd is running, which actually simplified the code in bridge-stp significantly.

commodo commented on bridge-stp.in:
```diff
@@ -29,34 +28,15 @@ net_dir=/sys/class/net
         if [ -e "$conf_file" ]; then
             . "$conf_file" || exit
         fi
-        checkpid $pid_file || exit 1
```
so much ack :)
thanks

commodo commented on bridge-stp.in:
```diff
@@ -28,31 +28,51 @@ net_dir=/sys/class/net
+# Determine whether mstpd should manage STP for the specified bridge.
+# Returns 0 if mstpd should manage STP for the specified bridge, or 1 if mstpd
+# should not manage STP for the specified bridge.
+is_mstp_bridge()
```
@PaulSD 
ack for this change too
question: are you going to re-spin that change regarding the "Coudln't connect to server" ? 

or are you dropping that change ?
fine either way from me

PaulSD replied:
I think I'll just drop that change.  Using pidof fixes the corner case I was trying to solve with the "Couldn't connect to server" thing.

commodo commented on bridge-stp.in:
```diff
@@ -62,18 +83,61 @@ case $2 in
+            # then dying before or when mstpctl connects to it.  To avoid that
+            # possibility, we instead simply turn STP off if `mstpctl addbridge`
+            # fails.
+            ( "$mstpctl" addbridge "$bridge" || brctl stp "$bridge" off ) &
```
i think there could be a better solution here that backgrounding this call
i'm a bit uncomfortable with this change ;

maybe we could think about another way of doing this
i'll try to think of one too

commodo commented on bridge-stp.in:
```diff
@@ -62,18 +83,61 @@ case $2 in
 
         # Add bridge to mstpd.
-        exec $bin_dir/mstpctl addbridge $bridge
+        "$mstpctl" addbridge "$bridge" || exit 3
```
this line is initially added in commit bdeae079aa38abdabb754a74d81046aec81fccae
maybe move this change there ?

PaulSD replied:
Ok, I'll move it.

commodo commented on bridge-stp.in:
```diff
@@ -62,18 +83,61 @@ case $2 in
     stop)
         # Remove bridge from mstpd.
-        exec $bin_dir/mstpctl delbridge $bridge
+        "$mstpctl" delbridge "$bridge" || exit 3
```
same for this line & commit bdeae079aa38abdabb754a74d81046aec81fccae

commodo commented on bridge-stp.in:
```diff
@@ -139,8 +164,59 @@ case $2 in
+
+            # Skip bridges that are not in user_stp mode.
+            if [ "$State" != '2' ]; then
+                echo
```
`echo >&2` ?

PaulSD replied:
Good catch.  Will fix.

PaulSD replied:
Actually, I think the echo lines after that should not have >&2 since they are informational messages, not errors.

commodo commented on bridge-stp.in:
```diff
@@ -139,8 +164,59 @@ case $2 in
+            # Skip bridges that have STP disabled.
+            read State < "$net_dir/$bridge/bridge/stp_state"
+            if [ "$State" = '0' ]; then
+                echo
```
echo >&2 ?

commodo commented on bridge-stp.in:
```diff
@@ -139,8 +164,59 @@ case $2 in
-        echo "Unknown action: $2" >&2
-        echo "Usage: bridge-stp <bridge> {start|stop}" >&2
-        exit 1
+    restart|restart_config)
```
ack

commodo commented on lib/mstpctl_restart_config:
```diff
@@ -1,96 +0,0 @@
-mstpctl_br_cfg      = ['maxage', 'fdelay', 'maxhops', 'txholdcount', \
-                      'forcevers', 'hello', 'ageing']
-mstpctl_br_msti_cfg = ['treeprio']
-mstpctl_prt_cfg      = ['portpathcost', 'portadminedge', 'portautoedge', \
```
and ack

commodo commented:
unless i've missed something, i think i've covered all commits

the only i'm uncertain about is this: 
https://github.com/mstpd/mstpd/pull/25/commits/c0e118834de25e44cf8f57ecb79b665d9e1e118c#diff-6aedd8ea85896ebe3d660cc5f4a17ac9R105

looking at it, it seems fairly contained in it's own case
in the sense, that it should only happen, if `MANAGE_MSTPD == y` and mstpd is not started by an init system

so, i'll propose this:
* i'm fine with merging this as is (rest commits are nitpicks anyway)
* we open a issue on this to look later ; 

i also want to extend more travis automation on the recent additions from Paul

commodo commented:
I forgot to add : if any objections please raise hands, or keys  @PaulSD @vivien @dvitas @asatish

PaulSD commented:
I've updated the commits per your comments.

Yeah, I agree that c0e1188#diff-6aedd8ea85896ebe3d660cc5f4a17ac9R105 is less than ideal, but I haven't been able to come up with any better way to deal with this, and this is only used in this one corner case.

commodo commented:
it took only 98 comments to get here :)
mine is 99th 

thanks :)

PaulSD commented:
Excellent!  Thanks so much for the careful review!
